src/fi_tostr.c

```diff
@@ -537,6 +537,102 @@ static void ofi_tostr_fabric_attr(char *buf, const struct fi_fabric_attr *attr,
 		FI_MAJOR(attr->api_version), FI_MINOR(attr->api_version));
 }
 
+static void ofi_tostr_device_attr(char *buf, const struct fi_device_attr *attr)
+{
+	const char *prefix = TAB TAB;
+	ofi_strcatf(buf, "%sfi_device_attr:\n", prefix);
+
+	prefix = TAB TAB TAB;
+	ofi_strcatf(buf, "%sname: %s\n", prefix, attr->name);
+	ofi_strcatf(buf, "%sdevice_id: %s\n", prefix, attr->device_id);
+	ofi_strcatf(buf, "%sdevice_version: %s\n", prefix, attr->device_version);
+	ofi_strcatf(buf, "%svendor_id: %s\n", prefix, attr->vendor_id);
+	ofi_strcatf(buf, "%sdriver: %s\n", prefix, attr->driver);
+	ofi_strcatf(buf, "%sfirmware: %s\n", prefix, attr->firmware);
+}
+
+static void ofi_tostr_pci_attr(char *buf, const struct fi_pci_attr *attr)
+{
+	const char *prefix = TAB TAB TAB;
+	ofi_strcatf(buf, "%sfi_pci_attr:\n", prefix);
+
+	prefix = TAB TAB TAB TAB;
+	ofi_strcatf(buf, "%sdomain_id: %u\n", prefix, attr->domain_id);
+	ofi_strcatf(buf, "%sbus_id: %u\n", prefix, attr->bus_id);
+	ofi_strcatf(buf, "%sdevice_id: %u\n", prefix, attr->device_id);
+	ofi_strcatf(buf, "%sfunction_id: %u\n", prefix, attr->function_id);
+}
+
+static void ofi_tostr_bus_type(char *buf, int type)
+{
+	switch (type) {
+	CASEENUMSTR(FI_BUS_UNKNOWN);
+	CASEENUMSTR(FI_BUS_PCI);
+	default:
+		ofi_strcatf(buf, "Unknown");
+		break;
+	}
+}
+
+static void ofi_tostr_bus_attr(char *buf, const struct fi_bus_attr *attr)
+{
+	const char *prefix = TAB TAB;
+	ofi_strcatf(buf, "%sfi_bus_attr:\n", prefix);
+
+	prefix = TAB TAB TAB;
+	ofi_strcatf(buf, "%sfi_bus_type: ", prefix);
+	ofi_tostr_bus_type(buf, attr->bus_type);
+	ofi_strcatf(buf, "\n");
+
+	switch (attr->bus_type) {
+	case FI_BUS_PCI:
+		ofi_tostr_pci_attr(buf, &attr->attr.pci);
+		break;
+	default:
+		break;
+	}
+}
+
+static void ofi_tostr_link_state(char *buf, int state)
+{
+	switch (state) {
+	CASEENUMSTR(FI_LINK_UNKNOWN);
+	CASEENUMSTR(FI_LINK_DOWN);
+	CASEENUMSTR(FI_LINK_UP);
```

Comment:
Shouldn't LINK_UP/DOWN output their states?

Reply:
This will output "FI_LINK_UP" or "FI_LINK_DOWN" as the state